packages/auth/amplify_auth_cognito_dart/lib/src/util/authorization_providers.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+@internal
+library amplify_auth_cognito.util.authorization_providers;
+
+import 'dart:async';
+
+import 'package:amplify_auth_cognito_dart/amplify_auth_cognito_dart.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:aws_signature_v4/aws_signature_v4.dart';
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+class AWSIamAuthProvider extends AmplifyAuthProvider {
+  @override
+  Future<http.BaseRequest> authorizeRequest(http.BaseRequest request,
+      {HttpRequestTransformOptions? options}) async {
+    if (options == null) {
+      throw const AuthException(
+        'Unable to authorize request with IAM. No region or service provided.',
+      );
+    }
+    final authSession = await Amplify.Auth.fetchAuthSession(
+            options: const CognitoSessionOptions(getAWSCredentials: true))
+        as CognitoAuthSession;
+    final credentials = authSession.credentials;
+    if (credentials == null) {
+      throw const ApiException(
```

Comment:
shouldn't be API error, but general callout to rethink errors from auth providers, whether needs special exception types, etc...

Reply:
I would say, in general, the exception type should not matter to the caller, meaning all exceptions should be handled equally by the caller. The specific type thrown should be native to the plugin in which it's defined, though, so an `AuthException` or `InvalidCredentialsException` from sigv4 package.